### PR TITLE
Support for Anchor Types in the Rust renderer

### PR DESCRIPTION
**Pull request by blockiosaurus**
With Anchor 0.30.0, IDLs are more dynamically generated using the IDL build trait. This is itself embedded in the AnchorSerialize trait. This PR providers a choice between Anchor and Borsh serialization to make it easier to interface between Anchor and MPL programs.

### CODE REVIEW

File: test/packages/rust/src/generated/types/rule_set.rs [hunk 1]
```diff
@@ -5,12 +5,17 @@
...
-#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(
+    not(feature = "anchor"),
+    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
+)]
+#[cfg_attr(
+    feature = "anchor",
+    derive(anchor_lang::AnchorSerialize, anchor_lang::AnchorDeserialize)
+)]
```
Comment (danenbm) [inline test/packages/rust/src/generated/types/rule_set.rs:18]: Looks like this lost the serde serialization on types.

Comment (blockiosaurus) [reply test/packages/rust/src/generated/types/rule_set.rs]: Fixed!

**Comment by lorisleiva** [comment]
Hey Keith, thanks for the PR! That makes sense. Do you know if that would be an issue with types imported from the `kaigan` crate? I'm not too sure but I think these only provide implementations for Borsh traits. Might be worth adding Anchor trait implementations in Kaigan too?

**Comment by blockiosaurus** [comment]
Nope, we've been testing this on Core. The traits are nearly identical because Anchor serde is basically Borsh anyway.

**Comment by febo** [comment]
@blockiosaurus Nice! Could you please add the following to `./test/packages/rust/Cargo.toml`?
1. `anchor = ["dep:anchor-lang"]` under the `[features]` section.
2. `anchor-lang = { version = "0.30.0", optional = true }` in the `[dependencies]` section.

If I understood correctly, this is needed to use the generated types on Anchor structs, so the Anchor IDL generation works. We might need to add something similar to Kaigan.

Core does not use Kaigan types, so it won't show whether there is a problem or not. Any chance you can try on Bubblegum to see what happens?

**Comment by blockiosaurus** [comment]
Thanks! Tracking that in a separate issue.

File: src/renderers/rust/templates/instructionsPage.njk [hunk 2]
```diff
@@ -129,8 +129,10 @@ impl {{ instruction.name | pascalCase }}InstructionData {
 }
 
 {% if hasArgs %}
-#[derive(BorshSerialize, BorshDeserialize, Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(not(feature = "anchor"), derive(BorshSerialize, BorshDeserialize))]
```
Comment (danenbm) [inline src/renderers/rust/templates/instructionsPage.njk:132]: same comment as other spot

**Review by danenbm — APPROVED** [review]
Approved with couple nit comments

**Comment by febo** [comment]
FYI - using the new `idl-build` feature with Kaigan types currently does not work.

image

**Comment by blockiosaurus** [comment]
Yup, tracking it as mentioned in the previous comment.

**Review by lorisleiva — APPROVED** [review]
Hey Keith, thanks again for this PR. I'm happy to merge this when you think it is ready. Is the plan to merge this, update `kaigan` and then publish the new version?

File: test/pubkey_tests.json [hunk 3]
```diff
@@ -52,7 +52,7 @@
   ],
   "metadata": {
     "origin": "shank",
-    "address": "PubkeyTests1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d",
+    "address": "PubkeyTests1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7",
```
Comment (lorisleiva) [inline test/pubkey_tests.json:55]: nit: What's the reason for this change?

Comment (blockiosaurus) [reply test/pubkey_tests.json:55]: The previous Pubkey was invalid.

**Comment by blockiosaurus** [comment]
Whoops just saw your message. Kinobi is published, but kaigan is pretty messy so I'm planning on doing that in a follow up release this sprint or next.
